File: okoshko.c
#define OKO_TEMP_ALLOCATOR_IMPLEMENTATION
#include "okoshko.h"

#include <stdint.h>
#include <stdlib.h>
#include <string.h>

// TODO: use oko_error everywhere

static oko_temp_allocator ta;
const char* oko_error = NULL;
const char* oko_error2 = NULL;

OKO_API void oko_init() { ta = oko_temp_init(64 * 1024 * 1024); }

#ifdef OKO_APPLE
#include "platform/apple.h"
#elif defined(OKO_WINDOWS)
#include "platform/windows.h"
#elif defined(OKO_LINUX)
#include "platform/linux_x11.h"
#endif

OKO_API void oko_set_fps(oko_Window* win, u32 fps) {
    if (fps <= 0)
    {
        win->vsync = false;
        win->target_frame_time = 0;
    }
    else
    {
        win->vsync = true;
        win->target_frame_time = (u64)((1000000.0 / fps) + 0.5);
    }
}

OKO_API u32 oko_get_fps(oko_Window* win) {
    if (win->actual_frame_time <= 0)
        return 0;
    return (u32)((1000000.0 / (f64)win->actual_frame_time) + 0.5);
}

OKO_API u8 oko_is_running(const oko_Window* win) { return win->running; }

OKO_API void oko_begin_drawing(oko_Window* win) {
    win->frame_start_time = okoshko_timer_now(win->timer);
    oko_poll_events(win);
}

OKO_API void oko_end_drawing(oko_Window* win) {
    memcpy(win->pixels, win->back_buffer, win->width * win->height * sizeof(u32));

#ifdef OKO_WINDOWS
    HDC hdc = GetDC(win->pw->hwnd);
    BITMAPINFO bmi = {0};
    bmi.bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
    bmi.bmiHeader.biWidth = win->width;
    bmi.bmiHeader.biHeight = -win->height;
    bmi.bmiHeader.biPlanes = 1;
    bmi.bmiHeader.biBitCount = 32;
    bmi.bmiHeader.biCompression = BI_RGB;
    StretchDIBits(hdc, 0, 0, win->width, win->height, 0, 0, win->width,
                  win->height, win->pixels, &bmi, DIB_RGB_COLORS, SRCCOPY);
    ReleaseDC(win->pw->hwnd, hdc);
#elif defined(OKO_LINUX)
    XPutImage(win->pw->dpy, win->pw->w, win->pw->gc, win->pw->img, 0, 0, 0, 0,
              win->width, win->height);
    XFlush(win->pw->dpy);
#endif

    u64 frame_end_time = okoshko_timer_now(win->timer);
    u64 elapsed_us = (frame_end_time - win->frame_start_time) * 1000;

    if (win->target_frame_time > 0)
    {
        u64 spin_threshold = 1000;

        if (elapsed_us + spin_threshold < win->target_frame_time)
        {
            u64 sleep_time_ms =
                (win->target_frame_time - elapsed_us - spin_threshold) / 1000;
            if (sleep_time_ms > 0)
            {
                okoshko_timer_sleep(sleep_time_ms);
            }
        }

        while (1)
        {
            frame_end_time = okoshko_timer_now(win->timer);
            elapsed_us = (frame_end_time - win->frame_start_time) * 1000;
            if (elapsed_us >= win->target_frame_time)
            {
                break;
            }
        }
    }

    win->actual_frame_time = elapsed_us;
    if (win->actual_frame_time < 1)
    {
        win->actual_frame_time = 1;
    }
}

OKO_API void oko_clear(oko_Window* win, u32 color) {
    u32* buf = win->back_buffer;
    i32 count = win->width * win->height;
    for (i32 i = 0; i < count; i++)
    {
        buf[i] = color;
    }
}

OKO_API void oko_set_pixel(oko_Window* win, i32 x, i32 y, u32 color) {
    if (x >= 0 && x < win->width && y >= 0 && y < win->height)
    {
        win->back_buffer[y * win->width + x] = color;
    }
}

OKO_API u32 oko_get_pixel(oko_Window* win, i32 x, i32 y) {
    if (x >= 0 && x < win->width && y >= 0 && y < win->height)
    {
        return win->back_buffer[y * win->width + x];
    }
    return 0;
}

OKO_API void oko_fill_rect(oko_Window* win, oko_Rect rect, u32 color) {
    i32 x0 = rect.x < 0 ? 0 : rect.x;
    i32 y0 = rect.y < 0 ? 0 : rect.y;
    i32 x1 = rect.x + rect.w > win->width ? win->width : rect.x + rect.w;
    i32 y1 = rect.y + rect.h > win->height ? win->height : rect.y + rect.h;

    for (i32 y = y0; y < y1; y++)
    {
        for (i32 x = x0; x < x1; x++)
        {
            win->back_buffer[y * win->width + x] = color;
        }
    }
}

OKO_API void oko_fill_circle(oko_Window* win, i32 cx, i32 cy, i32 radius,
                             u32 color) {
    for (i32 y = -radius; y <= radius; y++)
    {
        for (i32 x = -radius; x <= radius; x++)
        {
            if (x * x + y * y <= radius * radius)
            {
                oko_set_pixel(win, cx + x, cy + y, color);
            }
        }
    }
}

OKO_API void oko_draw_rect(oko_Window* win, oko_Rect rect, u32 color) {
    oko_draw_line(win, rect.x, rect.y, rect.x + rect.w, rect.y, color);
    oko_draw_line(win, rect.x + rect.w, rect.y, rect.x + rect.w, rect.y + rect.h,
                  color);
    oko_draw_line(win, rect.x + rect.w, rect.y + rect.h, rect.x, rect.y + rect.h,
                  color);
    oko_draw_line(win, rect.x, rect.y + rect.h, rect.x, rect.y, color);
}

OKO_API void oko_draw_line(oko_Window* win, i32 x0, i32 y0, i32 x1, i32 y1,
                           u32 color) {
    i32 dx = abs(x1 - x0);
    i32 dy = abs(y1 - y0);
    i32 sx = x0 < x1 ? 1 : -1;
    i32 sy = y0 < y1 ? 1 : -1;
    i32 err = dx - dy;

    while (1)
    {
        oko_set_pixel(win, x0, y0, color);
        if (x0 == x1 && y0 == y1)
            break;
        i32 e2 = 2 * err;
        if (e2 > -dy)
        {
            err -= dy;
            x0 += sx;
        }
        if (e2 < dx)
        {
            err += dx;
            y0 += sy;
        }
    }
}

OKO_API void oko_draw_circle(oko_Window* win, i32 cx, i32 cy, i32 radius,
                             u32 color) {
    i32 x = 0;
    i32 y = radius;
    i32 d = 3 - 2 * radius;

    while (y >= x)
    {
        oko_set_pixel(win, cx + x, cy + y, color);
        oko_set_pixel(win, cx - x, cy + y, color);
        oko_set_pixel(win, cx + x, cy - y, color);
        oko_set_pixel(win, cx - x, cy - y, color);
        oko_set_pixel(win, cx + y, cy + x, color);
        oko_set_pixel(win, cx - y, cy + x, color);
        oko_set_pixel(win, cx + y, cy - x, color);
        oko_set_pixel(win, cx - y, cy - x, color);
        x++;
        if (d > 0)
        {
            y--;
            d = d + 4 * (x - y) + 10;
        }
        else
        {
            d = d + 4 * x + 6;
        }
    }
}

OKO_API oko_Glyph oko_create_glyph(u8** bitmap, i32 width, i32 height,
                                   i32 startX, u8 character) {
    oko_Glyph glyph;
    glyph.character = character;
    glyph.width = width;
    glyph.height = height;
    glyph.advance = width;
    glyph.offsetX = 0;
    glyph.offsetY = 0;

    glyph.bitmap = (unsigned char*)malloc(width * height);
    if (glyph.bitmap)
    {
        for (int y = 0; y < height; y++)
        {
            memcpy(&glyph.bitmap[y * width], &bitmap[y][startX], width);
        }
    }

    return glyph;
}

OKO_API oko_Font* oko_bitmap_to_font(u8** bitmap, i32 totalWidth,
                                     i32 totalHeight, i32 glyphWidth,
                                     i32 glyphCount, u8 startChar) {
    if (!bitmap || totalWidth <= 0 || totalHeight <= 0 || glyphWidth <= 0 ||
        glyphCount <= 0)
    {
        return NULL;
    }

    oko_Font* font = (oko_Font*)malloc(sizeof(oko_Font));
    if (!font)
        return NULL;

    font->size = totalHeight;
    font->ascent = (int)(totalHeight * 0.8);
    font->descent = totalHeight - font->ascent;
    font->lineGap = (int)(totalHeight * 0.2);
    font->glyphCount = glyphCount;

    font->glyphs = (oko_Glyph*)malloc(sizeof(oko_Glyph) * glyphCount);
    if (!font->glyphs)
    {
        free(font);
        return NULL;
    }

    for (int i = 0; i < glyphCount; i++)
    {
        int startX = i * glyphWidth;
        font->glyphs[i] = oko_create_glyph(bitmap, glyphWidth, totalHeight, startX,
                                           startChar + i);
    }

    return font;
}

OKO_API void oko_free_font(oko_Font* font) {
    if (!font)
        return;

    if (font->glyphs)
    {
        for (int i = 0; i < font->glyphCount; i++)
        {
            free(font->glyphs[i].bitmap);
        }
        free(font->glyphs);
    }
    free(font);
}

OKO_API void oko_draw_text(oko_Window* win, const char* text, oko_Font* font,
                           i32 x, i32 y, float scale, u32 color) {
    if (!win || !text || !font || !font->glyphs)
        return;

    i32 cursorX = x;
    i32 cursorY = y;

    u8 r = (color >> 16) & 0xFF;
    u8 g = (color >> 8) & 0xFF;
    u8 b = color & 0xFF;

    for (const char* c = text; *c != '\0'; c++)
    {
        if (*c == '\n')
        {
            cursorX = x;
            cursorY += (i32)((font->size + font->lineGap) * scale);
            continue;
        }

        oko_Glyph* glyph = NULL;
        for (i32 i = 0; i < font->glyphCount; i++)
        {
            if (font->glyphs[i].character == *c)
            {
                glyph = &font->glyphs[i];
                break;
            }
        }

        if (!glyph || !glyph->bitmap)
        {
            cursorX += (i32)((f32)font->size * 0.5f * scale);
            continue;
        }

        i32 scaledWidth = (i32)(glyph->width * scale);
        i32 scaledHeight = (i32)(glyph->height * scale);

        for (i32 gy = 0; gy < scaledHeight; gy++)
        {
            for (i32 gx = 0; gx < scaledWidth; gx++)
            {
                i32 srcX = (i32)(gx / scale);
                i32 srcY = (i32)(gy / scale);

                if (srcX >= glyph->width || srcY >= glyph->height)
                    continue;

                u8 alpha = glyph->bitmap[srcY * glyph->width + srcX];

                if (alpha == 0)
                    continue;

                i32 px = cursorX + gx + (i32)(glyph->offsetX * scale);
                i32 py = cursorY + gy + (i32)(glyph->offsetY * scale);

                if (px < 0 || px >= win->width || py < 0 || py >= win->height)
                    continue;

                if (alpha == 255)
                {
                    win->back_buffer[py * win->width + px] = color;
                }
                else
                {
                    u32 bg = win->back_buffer[py * win->width + px];
                    u8 bgR = (bg >> 16) & 0xFF;
                    u8 bgG = (bg >> 8) & 0xFF;
                    u8 bgB = bg & 0xFF;

                    float a = alpha / 255.0f;
                    u8 outR = (u8)(r * a + bgR * (1.0f - a));
                    u8 outG = (u8)(g * a + bgG * (1.0f - a));
                    u8 outB = (u8)(b * a + bgB * (1.0f - a));

                    win->back_buffer[py * win->width + px] =
                        (outR << 16) | (outG << 8) | outB;
                }
            }
        }

        cursorX += (i32)(glyph->advance * scale);
    }
}

OKO_API u8 oko_key_down(oko_Window* win, u8 key) {
    return win->keyboard.keys[key];
}

OKO_API u8 oko_key_pressed(oko_Window* win, u8 key) {
    if (key >= 'A' && key <= 'Z')
    {
        if (win->keyboard.shift)
        {
            key = key + 32;
        }
    }
    return win->keyboard.keys[key];
}

OKO_API u8 oko_mouse_down(oko_Window* win, u8 button) {
    if (button == OKO_MOUSE_LEFT)
        return win->mouse.left;
    if (button == OKO_MOUSE_RIGHT)
        return win->mouse.right;
    if (button == OKO_MOUSE_MIDDLE)
        return win->mouse.middle;
    return 0;
}

OKO_API u64 oko_time_ms(oko_Window* win) {
    return okoshko_timer_now(win->timer);
}

OKO_API void oko_sleep(u64 ms) { okoshko_timer_sleep(ms); }

OKO_API char* oko_format(const char* format, ...) {
    va_list args;
    va_start(args, format);
    char* str = oko_temp_alloc(&ta, 512, 8);
    if (!str)
        return NULL;

    if (!vsprintf(str, format, args))
    {
        return NULL;
    }
    va_end(args);
    return str;
}

OKO_API oko_temp_allocator* oko_get_temp_allocator() { return &ta; }

OKO_API oko_AudioSystem* oko_audio_create() {
    oko_AudioSystem* audio = malloc(sizeof(oko_AudioSystem));
    if (!audio)
    {
        oko_error = "Could not allocate memory for audio system";
        return NULL;
    }
    audio->os_audio =
        oko_os_audio_create(OKO_AUDIO_SAMPLE_RATE, OKO_AUDIO_BUFFER_SIZE);
    if (!audio->os_audio)
    {
        oko_error = "Could not create os audio system";
        return NULL;
    }
    audio->is_open = 0;
    audio->is_paused = 1;
    audio->volume = 0.f;
    return audio;
}

OKO_API void oko_audio_destroy(oko_AudioSystem* audio) {
    oko_os_audio_destroy(audio->os_audio);
    free(audio);
}

OKO_API u64 oko_audio_get_available_frames(oko_AudioSystem* audio) {
    return oko_os_audio_get_available_frames(audio->os_audio);
}

OKO_API i32 oko_audio_write(oko_AudioSystem* audio, const f32* samples,
                            u64 frame_count) {
    return oko_os_audio_submit_buffer(audio->os_audio, samples, frame_count);
}

File: okoshko.zig
// TODO: proper wrapper not just binding

const std = @import("std");

pub const c = @cImport({
    @cInclude("./okoshko.h");
});

pub const Rect = extern struct {
    x: i32,
    y: i32,
    w: i32,
    h: i32,
};

pub const Point = extern struct {
    x: i32,
    y: i32,
};

pub const Keyboard = extern struct {
    keys: [256]u8,
    prev_keys: [256]u8,
    ctrl: u8,
    shift: u8,
    alt: u8,
    meta: u8,
};

pub const Mouse = extern struct {
    x: i32,
    y: i32,
    left: u8,
    right: u8,
    middle: u8,
    scroll_x: i32,
    scroll_y: i32,
};

pub const Glyph = extern struct {
    character: u8,
    width: i32,
    height: i32,
    advance: i32,
    offsetX: i32,
    offsetY: i32,
    bitmap: ?[*]u8,
};

pub const Font = extern struct {
    size: i32,
    ascent: i32,
    descent: i32,
    lineGap: i32,
    glyphCount: i32,
    glyphs: ?[*]Glyph,
};

pub const EventType = enum(c_int) {
    no_event = c.OKO_NO_EVENT,
};

pub const EventCallback = *const fn (?*anyopaque) callconv(.C) void;

pub const Timer = opaque {};
pub const PlatformWindow = opaque {};

pub const Window = c.oko_Window;

pub const OkoError = error{
InitFailed,
CreateFailed,
InvalidWindow,
InvalidFont,
};

pub fn getLastError() ?[]const u8 {
    if (c.oko_error) |err| {
        return std.mem.span(err);
    }
    return null;
}

pub fn getLastError2() ?[]const u8 {
    if (c.oko_error2) |err| {
        return std.mem.span(err);
    }
    return null;
}

pub fn init() void {
    c.oko_init();
}

pub fn createWindow(title: [:0]const u8, width: i32, height: i32) OkoError!*Window {
    const win = c.oko_create(title.ptr, width, height);
    return win orelse OkoError.CreateFailed;
}

pub fn destroyWindow(win: *Window) void {
    c.oko_destroy(win);
}

pub fn setFps(win: *Window, fps: u32) void {
    c.oko_set_fps(win, fps);
}

pub fn getFps(win: *Window) u32 {
    return c.oko_get_fps(win);
}

pub fn isRunning(win: *const Window) bool {
    return c.oko_is_running(win) != 0;
}

pub fn createTimer() OkoError!*Timer {
    const timer = c.okoshko_timer_create();
    return timer orelse OkoError.InitFailed;
}

pub fn timerNow(timer: *Timer) u64 {
    return c.okoshko_timer_now(timer);
}

pub fn timerSleep(ms: u64) void {
    c.okoshko_timer_sleep(ms);
}

pub fn swapBuffers(win: *Window) void {
    c.oko_os_swap_buffers(win);
}

pub fn beginFrame(win: *Window) void {
    c.oko_begin_frame(win);
}

pub fn endFrame(win: *Window) void {
    c.oko_end_frame(win);
}

pub fn pollEvents(win: *Window) void {
    c.oko_poll_events(win);
}

pub fn clear(win: *Window, color: u32) void {
    c.oko_clear(win, color);
}

pub fn setPixel(win: *Window, x: i32, y: i32, color: u32) void {
    c.oko_set_pixel(win, x, y, color);
}

pub fn getPixel(win: *Window, x: i32, y: i32) u32 {
    return c.oko_get_pixel(win, x, y);
}

pub fn fillRect(win: *Window, rect: Rect, color: u32) void {
    c.oko_fill_rect(win, @bitCast(rect), color);
}

pub fn fillCircle(win: *Window, cx: i32, cy: i32, radius: i32, color: u32) void {
    c.oko_fill_circle(win, cx, cy, radius, color);
}

pub fn drawRect(win: *Window, rect: Rect, color: u32) void {
    c.oko_draw_rect(win, @bitCast(rect), color);
}

pub fn drawLine(win: *Window, x0: i32, y0: i32, x1: i32, y1: i32, color: u32) void {
    c.oko_draw_line(win, x0, y0, x1, y1, color);
}

pub fn drawCircle(win: *Window, cx: i32, cy: i32, radius: i32, color: u32) void {
    c.oko_draw_circle(win, cx, cy, radius, color);
}

pub fn drawText(win: *Window, text: [:0]const u8, font: *Font, x: i32, y: i32, scale: f32, color: u32) void {
    c.oko_draw_text(win, text.ptr, font, x, y, scale, color);
}

pub fn keyDown(win: *Window, key: u8) bool {
    return c.oko_key_down(win, key) != 0;
}

pub fn keyPressed(win: *Window, key: u8) bool {
    return c.oko_key_pressed(win, key) != 0;
}

pub fn keyReset(win: *Window, key: u8) void {
    c.oko_key_reset(win, key);
}

pub fn mouseDown(win: *Window, button: u8) bool {
    return c.oko_mouse_down(win, button) != 0;
}

pub fn timeMs(win: *Window) u64 {
    return c.oko_time_ms(win);
}

pub fn sleep(ms: u64) void {
    c.oko_sleep(ms);
}

pub fn createGlyph(bitmap: *[*]u8, width: i32, height: i32, startX: i32, character: u8) Glyph {
    return @bitCast(c.oko_create_glyph(bitmap, width, height, startX, character));
}

pub fn bitmapToFont(
    bitmap: *[*]u8,
    totalWidth: i32,
    totalHeight: i32,
    glyphWidth: i32,
    glyphCount: i32,
    startChar: u8,
) OkoError!*Font {
    const font = c.oko_bitmap_to_font(bitmap, totalWidth, totalHeight, glyphWidth, glyphCount, startChar);
    return font orelse OkoError.InvalidFont;
}

pub fn freeFont(font: *Font) void {
    c.oko_free_font(font);
}

pub fn format(allocator: std.mem.Allocator, comptime fmt: []const u8, args: anytype) ![]u8 {
    return std.fmt.allocPrint(allocator, fmt, args);
}

pub fn getTempAllocator() *c.oko_temp_allocator {
    return c.oko_get_temp_allocator();
}

pub const Color = struct {
    pub fn rgba(r: u8, g: u8, b: u8, a: u8) u32 {
        return (@as(u32, a) << 24) | (@as(u32, b) << 16) | (@as(u32, g) << 8) | @as(u32, r);
    }

    pub fn rgb(r: u8, g: u8, b: u8) u32 {
        return rgba(r, g, b, 255);
    }

    pub const BLACK = rgb(0, 0, 0);
    pub const WHITE = rgb(255, 255, 255);
    pub const RED = rgb(255, 0, 0);
    pub const GREEN = rgb(0, 255, 0);
    pub const BLUE = rgb(0, 0, 255);
    pub const YELLOW = rgb(255, 255, 0);
    pub const CYAN = rgb(0, 255, 255);
    pub const MAGENTA = rgb(255, 0, 255);
};